### DISCUSSION TITLE
Unexpectly bump golang version

### DISCUSSION THREAD

**Discussion by MuZhou233** [renovatebot/renovate#24994] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

In a recent PR that renovate created in my project, it update golang version from 1.20 to 1.21. But that not shown in updates list in the PR message body.

Please check https://github.com/tuihub/librarian/pull/72

My questions:
1. As I know, renovate won't bump golang version because it is disabled by default and I didn't enable it. Why this happened.
2. Why golang version info is missing in updates list.
3. How can I configure to let renovate bump golang version in a sepetared PR (note that my current config could be found at https://github.com/tuihub/librarian/blob/master/renovate.json).

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
I think that may be a result of `go get`, and if so then it's new behavior from Go. Can you test with a basic repo (e.g. one dependency, go 1.20) and see if it also happens?

**Reply by MuZhou233** [reply]
I have setup a test repo with similar configs. That didn't happen this time.

Check https://github.com/tuihub/go-test-repo/pull/1

**Reply by rarkins** [reply]
Please check and compare the logs for both. You want to look for when "docker run" happened on each

**Reply by MuZhou233** [reply]
I didn't find the point. Could you explain more?

**Reply by rarkins** [reply]
You need to look at developer.mend.io under each repo and find the job/log from when the commit was made

**Reply by MuZhou233** [reply]
FInally this indeed is a new behavor from Go 1.21. I prefer to call it a breaking change.
Now the go version will be auto upgrade if any dependency using 1.21. And that is easy to break old CI config.
I could temporarily fix my issue. But I think renovate should make progress on it :)

Please check https://github.com/golang/go/issues/57001 and https://github.com/golang/go/issues/62409

**Reply by rarkins** [reply]
But why do you think it didn't reproduce when you tried?

**Reply by MuZhou233** [reply]
It will happen when one dependency upgrade Go version to 1.21. My previous try didn't met this.

Minimal reproduction available now at https://github.com/tuihub/go-test-repo/pull/1

**Reply by rarkins** [reply]
OK so the choice is:
- Stay on same version of Go and fail
- Run with newer version of Go and it will self-updated the go.mod directive

Isn't the second better?

**Reply by rarkins** [reply]
Go does not seem to update the go.mod directive unless it's necessary. Example here of it not updating a 1.19 go.mod file: https://github.com/renovate-tests/column/pull/5/files

**Reply by MuZhou233** [reply]
I agree. There is another clarify here https://github.com/golang/go/issues/63279.

The point is, the default behaviour is different between old version and new of Go toolchan.
Before 1.21, use another module with a higher Go version as dependency is allowed, just like your example does. And after 1.21, it will force upgrade to the highest Go version among dependencies'.

I believe it's also related to https://github.com/renovatebot/renovate/issues/16715. Before, so many developers treat Go minor version as major version and avoid to automatically upgrade it. But now Go team is forcing us to make change.

It's quiet complex so I think invite more people to discuss can help us on it.

**Reply by MuZhou233** [reply]
My advice:

1. Implement the ability to detect Go version change and contain that info in PR body.
2. Provide a new way to leave Go version alone and block some dependencies' update which will influence the Go version(or provide ability to disable auto-merge when Go version change detected).
3. Discuss which is the default behaviour.

**Reply by rarkins** [reply, marked as answer]
For (1) I created #25047 for indicating that the Go.mod version has been implicitly bumped.

For (2) I created #25048 which would allow you to suppress such PRs if desired.

Finally I've created #25049 to give visibility into all updated direct dependencies
